### PR TITLE
Boilerplate server code

### PR DESCRIPTION
**Pull request by klanmiko**
This commit introduces a new folder structure

src contains all of the server code, it's entry point should be the containers/routes.py file which defines the REST API endpoints, called by app.py

src/db contains the files that manage the database connection and schema

src/db/connection manages the database connections, it introduces these requirements for the database:

- Regular user that has access to select and update rows
- Admin user for CSV uploading that has access to modify the Applications table columns using ALTER TABLE and possibly DROP TABLE

src/db/models defines classes that can be used to retrieve data from the database. Each instance holds a reference to the database connection and have easy member functions that can perform the database requests needed to fulfill a request. Possibly extend to implement a transactional model if needed

src/lib defines all helper scripts and non-core functionality

src/lib/loadCSV.py defines functions that serve the CSV upload request and structuring the database

src/controller this folder defines the REST API endpoint controller and manages serving views, manages session, performing auth, and performing operations on the database

src/controller/routes.py is empty currently but should hold the routes, when I figure out how to use flask. @thomasmunduchira can you hook this up with the app.py?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -103,6 +104,8 @@ venv.bak/
 # mypy
 .mypy_cache/
 
+# vscode
```
Comment (thomasmunduchira) [inline .gitignore]: Can you put editor-specific gitignore rules in your global gitignore?

File: requirements.txt [hunk 2]
```diff
@@ -1,6 +1,18 @@
+atomicwrites==1.1.5
```
Comment (thomasmunduchira) [inline requirements.txt]: Are all these python packages required? Looks they were generated by `pip freeze`.

File: src/db/connection.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+def getConnection():
```
Comment (thomasmunduchira) [inline src/db/connection.py]: Python style is to use camelCase, so can you make this get_connection?

File: src/db/models.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    def __init__(self, db):
+        """ store a database connection """
+        self._db = db
+
+    def store(self, row):
+        """ does an INSERT INTO with the data in the row """
+        pass
+
+
+def setApplicationsModel(fieldNames):
```
Comment (thomasmunduchira) [inline src/db/models.py]: set_applications_model(fieldNames):

File: src/db/models.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    """
+    sets up a global table for applications
+    
+    fieldNames -- a list of strings for the fields in the typeform CSV
+    """
+    pass
+
+def getApplicationsModel():
```
Comment (thomasmunduchira) [inline src/db/models.py]: get_applications_model():

File: src/lib/loadCSV.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+from ..db import models
+import csv
+
+def _storeCSVRow(row):
```
Comment (thomasmunduchira) [inline src/lib/loadCSV.py]: _store_csv_row(row):

File: src/lib/loadCSV.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    """ 
+    Read a line from the CSV and store it in the database using the model 
+
+    row -- a python dictionary defining the data in the row
+    """
+    applicationsEntry = models.getApplicationsModel()
+    applicationsEntry.store(row)
+
+def uploadCSV(csvFile):
```
Comment (thomasmunduchira) [inline src/lib/loadCSV.py]: upload_csv(csvFile):

File: start.py [hunk 8]
```diff
@@ -1,21 +1,45 @@
...
-    'install': "pip3 install -r requirements.txt && cd applications-frontend && npm install && cd ..",
+    'run-other': "export FLASK_APP=app.py && flask run",
+    'install': "pip3 install -r requirements.txt",
     'format': "autopep8 --in-place --aggressive --aggressive *.py",
-    'deploy': "pm2 start start.sh",
-    'publish-deps': 'pip3 freeze > requirements.txt'
-}
+    'deploy': "pm2 start deploy.sh",
+    'save-deps': 'pip3 freeze > requirements.txt',
+    'install-npm': "cd applications-frontend && npm install && cd .."}
```
Comment (thomasmunduchira) [inline start.py]: Can we call this `install-client` or something similar?

File: .gitignore (hunk 1 above)
Comment (thebearjew) [reply .gitignore]: This isn't global, and you should absolutely ignore any/all editor related files that are generated.

File: src/db/connection.py (hunk 3 above)
Comment (thebearjew) [reply src/db/connection.py]: That's not true.

```py
def getConnection(): # This is camelCase

def get_connection(): # This is called snake_case
```

Comment (thomasmunduchira) [reply src/db/connection.py]: Ah I messed that up. I did mean snake_case

File: .gitignore (hunk 1 above)
Comment (thomasmunduchira) [reply .gitignore]: From the docs on git (`https://git-scm.com/docs/gitignore`):
```
Patterns which a user wants Git to ignore in all situations (e.g., backup or temporary files generated by the user’s editor of choice) generally go into a file specified by core.excludesFile in the user’s ~/.gitconfig. Its default value is $XDG_CONFIG_HOME/git/ignore. If $XDG_CONFIG_HOME is either not set or empty, $HOME/.config/git/ignore is used instead.
```

I agree that editor-related files should be ignored, just that this isn't the right place for it by convention.

File: src/db/models.py (hunk 4 above)
Comment (klanmiko) [reply src/db/models.py]: --- fieldnames ---

File: src/lib/loadCSV.py (hunk 7 above)
Comment (klanmiko) [reply src/lib/loadCSV.py]: csv_file

File: start.py (hunk 8 above)
Comment (klanmiko) [reply start.py]: ok

File: requirements.txt (hunk 2 above)
Comment (klanmiko) [reply requirements.txt]: they were

File: .gitignore (hunk 1 above)
Comment (klanmiko) [reply .gitignore]: Good idea, still keeping this one line in here though, in this massively tiny file

**Pull request merged**